### ISSUE TITLE
Tesla binding going offline

### ISSUE THREAD

**Issue by hlemmens** [openhab/openhab-addons#14435]
Since about a month, my Tesla binding is going offline every 8 hours. It seems to be a problem of getting a new access token once the old one has expired. After manually disabling and enabling the binding, everything works fine for another 8 hours.

Anyone experiencing the same issue? Any solution?

**Comment by wborn** [comment]
I think it was fixed with https://github.com/openhab/openhab-addons/pull/14300?

**Comment by hlemmens** [comment]
Thanks for the hint. I’m currently running v 3.4.1 of the Tesla binding. Where to find the new version and how to install it? Can’t find it 🥴Regards,Hubert LemmensOp 18 feb. 2023 om 21:11 heeft Wouter Born ***@***.***> het volgende geschreven:﻿
I think it was fixed with #14300?

**Comment by wborn** [comment]
It depends on your installation method.

openHABian has a tool that for it:

https://www.openhab.org/docs/installation/openhabian.html#openhabian-configuration-tool

**Comment by hlemmens** [comment]
OK Wouter; thanks

I managed to install the version 3.4.2 of the Tesla binding. Hope this is the right one.

**Comment by wborn** [comment]
Does it fix your issue @hlemmens?

**Comment by hlemmens** [comment]
Unfortunately not. But I'm not sure the #14300 fix was implemented in v3.4.2. . How to figure out?

**Comment by wborn** [comment]
It should be part of it according to the release notes:

https://github.com/openhab/openhab-distro/releases/tag/3.4.2

**Comment by hlemmens** [comment]
OK. Thanks. There is another modification ongoing or finalised: #14351. I think it’s not yet in the v3.4.2. Maybe this one can solve the problem? Where to find the .jar file?

**Comment by hlemmens** [comment]
The version 3.4.2 doesn't solve this #14435 issue. Any other hints?

**Comment by wborn** [comment]
I don't use it myself, maybe @kaikreuzer or @kgoderis can help.

**Comment by kaikreuzer** [comment]
I can't. For me it is absolutely stable.
What might help could be some debug logs from the moment when it goes offline.

**Comment by hlemmens** [comment]
I don’t find any Tesla related messages neither in the openhab.log nor the event.log. But anyhow, 8 hours after re-enabling the bridge it went offline. Weird.

**Comment by lsiepel** [comment]
You need to enable DEBUG logging. Documentation is here: 
https://www.openhab.org/docs/administration/logging.html

TL;DR from the console: `log:set DEBUG org.openhab.binding.tesla`

**Comment by hlemmens** [comment]
OK; thanks for the help. Debug is on now. Wait another 8 hours to see what happens.

**Comment by hlemmens** [comment]
This is what I could read in the log file:
2023-02-23 20:00:09.271 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has expired, trying to get a new one.
2023-02-23 20:00:09.273 [DEBUG] [internal.handler.TeslaAccountHandler] - Current authentication time 2023-02-23 20:00:09
2023-02-23 20:00:09.275 [DEBUG] [internal.handler.TeslaAccountHandler] - Found a request token from 2023-02-23 11:59:18
2023-02-23 20:00:09.277 [DEBUG] [internal.handler.TeslaAccountHandler] - Access token expiration time 2023-02-23 19:59:18
2023-02-23 20:00:09.278 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has expired
2023-02-23 20:00:09.280 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - Exchanging SSO refresh token for API access token
2023-02-23 20:00:09.281 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has expired, trying to get a new one.
2023-02-23 20:00:09.284 [DEBUG] [internal.handler.TeslaAccountHandler] - Current authentication time 2023-02-23 20:00:09
2023-02-23 20:00:09.287 [DEBUG] [internal.handler.TeslaAccountHandler] - Found a request token from 2023-02-23 11:59:18
2023-02-23 20:00:09.290 [DEBUG] [internal.handler.TeslaAccountHandler] - Access token expiration time 2023-02-23 19:59:18
2023-02-23 20:00:09.292 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has expired
2023-02-23 20:00:09.294 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - Exchanging SSO refresh token for API access token
2023-02-23 20:00:09.314 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - An exception occurred while invoking a HTTP request: 'java.io.EOFException: HttpConnectionOverHTTP@1057eb::DecryptedEndPoint@1dda816{l=/192.168.0.181:57782,r=auth.tesla.com/23.55.96.81:443,OPEN,fill=-,flush=-,to=28850510/0}'
2023-02-23 20:00:09.316 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - An error occurred during refresh of SSO token: no response
2023-02-23 20:00:09.319 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_state: 401: Unauthorized
2023-02-23 20:00:09.915 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - Access token expires in 28800 seconds at 2023-02-24 04:00:09
2023-02-23 20:00:09.918 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request set_charging_amps: 401: Unauthorized
2023-02-23 20:00:12.168 [DEBUG] [internal.handler.TeslaVehicleHandler] - Car has not moved in 5 min, and can sleep
2023-02-23 20:00:12.170 [DEBUG] [internal.handler.TeslaVehicleHandler] - fastpoll: Throttled to allow sleep, occupied/idle, or in a console mode
2023-02-23 20:00:14.654 [DEBUG] [internal.handler.TeslaVehicleHandler] - Querying the vehicle, response : 200, OK
2023-02-23 20:00:14.655 [DEBUG] [internal.handler.TeslaVehicleHandler] - Querying the vehicle: VIN 5YJ3E7EB1KF240549


It seems the bridge had some trouble to get a fresh access token. After a few attempts, it succeeded. But often it fails and I have to disable the bridge and then enable it.

**Comment by lsiepel** [comment]
The interesting part:
`'java.io.EOFException: HttpConnectionOverHTTP@1057eb::DecryptedEndPoint@1dda816{l=/192.168.0.181:57782,r=auth.tesla.com/23.55.96.81:443,OPEN,fill=-,flush=-,to=28850510/0}'`

Without the request and response it is hard to say why this end of file exception occurs. The binding does not have a trace logs set to capture those details, so without an updated binding version it is hard to tell what is causing this. It might also be something on the Tesla side, hard to tell.

**Comment by oOMagicOo** [comment]
I have the same problem since a few month. Already surfaced the problem in the OH forum [here](https://community.openhab.org/t/tesla-items-not-updated-after-some-time/138374/29?page=2). Eventually I wrote a script which restarts the binding every 8hrs which is not nice but works for me.

**Comment by lsiepel** [comment]
> I have the same problem since a few month. Already surfaced the problem in the OH forum [here](https://community.openhab.org/t/tesla-items-not-updated-after-some-time/138374/29?page=2). Eventually I wrote a script which restarts the binding every 8hrs which is not nice but works for me.

I could write some workaround this weekend. That if a eof exception occurs it performs one or two retries. But that is just a workaround the real cause is not yet known. I might also improve the logging.

**Comment by hlemmens** [comment]
Looking into the debug logs, I noticed some commands are no longer accepted (start charging, set charging amps), although the access key has not yet expired. At the same moment, the querying the vehicle seems to work.

Here goes a subset of logs:
2023-02-25 12:52:09.255 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_state: 408: Request Timeout
2023-02-25 12:52:09.260 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request set_charging_amps: 408: Request Timeout
2023-02-25 12:52:09.299 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_start: 408: Request Timeout
2023-02-25 12:52:09.878 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_state
2023-02-25 12:52:10.179 [WARN ] [internal.handler.TeslaAccountHandler] - Reached the maximum number of errors (3) for the current interval (15 seconds)
2023-02-25 12:52:10.182 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_state: 408: Request Timeout
2023-02-25 12:52:23.516 [DEBUG] [internal.handler.TeslaAccountHandler] - Setting up an authenticated connection to the Tesla back-end
2023-02-25 12:52:23.518 [DEBUG] [internal.handler.TeslaAccountHandler] - Current authentication time 2023-02-25 12:52:23
2023-02-25 12:52:23.520 [DEBUG] [internal.handler.TeslaAccountHandler] - Found a request token from 2023-02-25 08:23:35
2023-02-25 12:52:23.522 [DEBUG] [internal.handler.TeslaAccountHandler] - Access token expiration time 2023-02-25 16:23:35
2023-02-25 12:52:23.524 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has not expired yet
2023-02-25 12:52:24.006 [DEBUG] [internal.handler.TeslaAccountHandler] - Querying the vehicle: Response: 200: OK
2023-02-25 12:52:24.009 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: vehicle_config
2023-02-25 12:52:24.252 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request vehicle_config: 408: Request Timeout
2023-02-25 12:52:24.263 [DEBUG] [internal.handler.TeslaAccountHandler] - Retrying to send the command set_charging_amps.
2023-02-25 12:52:24.265 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: set_charging_amps
2023-02-25 12:52:24.302 [DEBUG] [internal.handler.TeslaAccountHandler] - Retrying to send the command charge_start.
2023-02-25 12:52:24.304 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_start
2023-02-25 12:52:24.547 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request set_charging_amps: 408: Request Timeout
2023-02-25 12:52:24.622 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_start: 408: Request Timeout
2023-02-25 12:52:33.370 [DEBUG] [internal.handler.TeslaVehicleHandler] - Querying the vehicle, response : 200, OK
2023-02-25 12:52:33.371 [DEBUG] [internal.handler.TeslaVehicleHandler] - Querying the vehicle: VIN 5YJ3E7EB1KF240549
2023-02-25 12:52:33.373 [DEBUG] [internal.handler.TeslaVehicleHandler] - Vehicle state is asleep
2023-02-25 12:52:33.374 [DEBUG] [internal.handler.TeslaVehicleHandler] - Vehicle is asleep.
2023-02-25 12:52:39.549 [DEBUG] [internal.handler.TeslaAccountHandler] - Retrying to send the command set_charging_amps.
2023-02-25 12:52:39.551 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: set_charging_amps
2023-02-25 12:52:39.623 [DEBUG] [internal.handler.TeslaAccountHandler] - Retrying to send the command charge_start.
2023-02-25 12:52:39.625 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_start
2023-02-25 12:52:39.848 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request set_charging_amps: 408: Request Timeout
2023-02-25 12:52:39.878 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_start: 408: Request Timeout

**Comment by hlemmens** [comment]
Another snippet of logs shows that when the vehicle is not waked up, the charge-start command and set charging amps command fails. After waking up, it works well.

 Retrying to send the command set_charging_amps.
2023-02-25 13:02:54.932 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: set_charging_amps
2023-02-25 13:02:55.099 [DEBUG] [internal.handler.TeslaAccountHandler] - Retrying to send the command charge_start.
2023-02-25 13:02:55.101 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_start
2023-02-25 13:02:55.311 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request set_charging_amps: 408: Request Timeout
2023-02-25 13:02:55.372 [DEBUG] [internal.handler.TeslaAccountHandler] - An error occurred while communicating with the vehicle during request charge_start: 408: Request Timeout
2023-02-25 13:03:08.864 [DEBUG] [internal.handler.TeslaVehicleHandler] - handleCommand tesla:model3:b31e2fa7ff:5YJ3E7EB1KF240549:charge ON
2023-02-25 13:03:08.866 [DEBUG] [internal.handler.TeslaVehicleHandler] - Waking vehicle to send command.
2023-02-25 13:03:08.869 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: wake_up
2023-02-25 13:03:08.869 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_start
2023-02-25 13:03:08.870 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: charge_state
2023-02-25 13:03:08.890 [DEBUG] [internal.handler.TeslaVehicleHandler] - handleCommand tesla:model3:b31e2fa7ff:5YJ3E7EB1KF240549:chargingamps 18 A
2023-02-25 13:03:08.895 [DEBUG] [internal.handler.TeslaVehicleHandler] - Waking vehicle to send command.
2023-02-25 13:03:08.896 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: wake_up
2023-02-25 13:03:08.896 [DEBUG] [internal.handler.TeslaAccountHandler] - Invoking: set_charging_amps
2023-02-25 13:03:09.328 [DEBUG] [internal.handler.TeslaVehicleHandler] - The request (charge_start) execution was successful, and reported ''
2023-02-25 13:03:09.351 [DEBUG] [internal.handler.TeslaVehicleHandler] - The request (set_charging_amps) execution was successful, and reported ''

**Comment by hlemmens** [comment]
It seems to be related to the vehicle state sleeping. the wake-up command doesn't always succeed.

**Comment by lsiepel** [comment]
> It seems to be related to the vehicle state sleeping. the wake-up command doesn't always succeed.

But restarting the binding always works?

**Comment by hlemmens** [comment]
Latest logs: no reply on the request for a new access token.

2023-02-25 16:23:35.355 [DEBUG] [internal.handler.TeslaVehicleHandler] - Querying the vehicle, response : 401, Unauthorized
2023-02-25 16:23:35.356 [DEBUG] [internal.handler.TeslaVehicleHandler] - The access token has expired, trying to get a new one.
2023-02-25 16:23:35.358 [DEBUG] [internal.handler.TeslaAccountHandler] - Current authentication time 2023-02-25 16:23:35
2023-02-25 16:23:35.359 [DEBUG] [internal.handler.TeslaAccountHandler] - Found a request token from 2023-02-25 08:23:35
2023-02-25 16:23:35.360 [DEBUG] [internal.handler.TeslaAccountHandler] - Access token expiration time 2023-02-25 16:23:35
2023-02-25 16:23:35.361 [DEBUG] [internal.handler.TeslaAccountHandler] - The access token has expired
2023-02-25 16:23:35.362 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - Exchanging SSO refresh token for API access token
2023-02-25 16:23:35.387 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - An exception occurred while invoking a HTTP request: 'java.io.EOFException: HttpConnectionOverHTTP@1c20e06::DecryptedEndPoint@7d0c51{l=/192.168.0.181:37224,r=auth.tesla.com/2.21.16.79:443,OPEN,fill=-,flush=-,to=28800190/0}'
2023-02-25 16:23:35.388 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - An error occurred during refresh of SSO token: no response
2023-02-25 16:23:35.389 [DEBUG] [internal.handler.TeslaVehicleHandler] - An error occurred while querying the vehicle

**Comment by lsiepel** [comment]
Maybe waking the car prior to refreshing the token maight help, but i'm to unfamilair with the API and i don;t own such a car, so i think @kaikreuzer  or @kgoderis might be of more help to proceed with these logs and a workaround.

**Comment by hlemmens** [comment]
> I have the same problem since a few month. Already surfaced the problem in the OH forum [here](https://community.openhab.org/t/tesla-items-not-updated-after-some-time/138374/29?page=2). Eventually I wrote a script which restarts the binding every 8hrs which is not nice but works for me.

Would you be so kind to share your script with me?

**Comment by lavanden** [comment]
Same issue here.

Since a restart of the binding always works, could it be that the httpClient is acting up? 

What if you pass in the httpClientFactory to the TeslaSSOHandler and getting a httpClient every time you call getAccessToken?

**Comment by lavanden** [comment]
@kaikreuzer, @kgoderis ,

Following changes seem to work for me:

in TeslaSSOHandler, in method getAccessToken, create a new httpClient for each call
`
HttpClient httpClient = httpClientFactory.createHttpClient("teslaSSOClient");
if (!httpClient.isStarted()) {
    try {
        httpClient.start();
    } catch (Exception e) {
        logger.debug("Exception while starting httpClient");
    }
}
final org.eclipse.jetty.client.api.Request request = httpClient.newRequest(URI_SSO + "/" + PATH_TOKEN);

request.content(new StringContentProvider(refreshTokenPayload));
request.header(HttpHeader.CONTENT_TYPE, "application/json");
request.method(HttpMethod.POST);

ContentResponse refreshResponse = executeHttpRequest(request);

if (httpClient.isStarted()) {
    try {
        httpClient.stop();
        httpClient.destroy();
    } catch (Exception e) {
        logger.debug("Exception while starting httpClient");
    }
}
`

However, I don't know if creating clients every call is a good idea. There might be better ways but I don't know since I'm not a Java/OpenHab developer.

**Comment by thijslemmens** [comment]
@lavanden  I implemented your fix for 3.4.4:
https://drive.google.com/file/d/13ntyqtyML22IOovsJbhVaHSTdibzbuUF/view?usp=sharing

This is the commit: https://github.com/thijslemmens/openhab-addons/commit/851d5691215e25158d8c58b15142408dcd2dd596

I can make an official PR, but I don't know if it has a chance of getting accepted.

**Comment by lsiepel** [comment]
@thijslemmens It probably only hides the real problem. Switching from the common http client to a 'custom' client is not something i would advice. We should find out why the  clients needs to be recreated. What property/circumstances is causing this behaviour with the client that it needs to be restarted?

**Comment by thijslemmens** [comment]
> @thijslemmens It probably only hides the real problem. Switching from the common http client to a 'custom' client is not something i would advice. We should find out why the  clients needs to be recreated. What property/circumstances is causing this behaviour with the client that it needs to be restarted?
> 

I agree. I built this version for @hlemmens so he does not need to manually restart the binding every time.
Any suggestions of what we can try to get closer to the real problem? I can build a version with extra debugging, but I'm not sure what to try next.

**Comment by lavanden** [comment]
@lsiepel ,

I tried debugging the issue,  but I could not get to the root. I'm thinking that the connection with the tesla backend gets dropped after 28800 seconds.

**Comment by lsiepel** [comment]
Previously your logs recorded:

> 2023-02-25 16:23:35.387 [DEBUG] [sla.internal.handler.TeslaSSOHandler] - An exception occurred while invoking a HTTP request: 'java.io.EOFException: HttpConnectionOverHTTP@1c20e06::DecryptedEndPoint@7d0c51{l=/192.168.0.181:37224,r=auth.tesla.com/2.21.16.79:443,OPEN,fill=-,flush=-,to=28800190/0}'

You could try to log the raw request/response to a file and see how the first (binding init) and last (when it crashed) differ. 

If that leads to nothing, you might also handle the EOFException by retrying / resetting all tokens and retry or something. Is that of any help? I don;t own a Teslo so i can't test, but this would be my aproach i guess.

**Comment by lsiepel** [comment]
Some connection related fixes have been merged after 3.4.4. Not sure if it also affects this issue, But could be usefull to check latest 4.1 snapshot or mile stone.

**Comment by Rucksackgesicht** [comment]
Hi guys,
have trouble with tesla binding 4.1.1. My account is offline so I can't control anything. It tries to go online, every 15 seconds but switches to offline immediately. I just did an openhab upgrade to 4.1.1 on sunday. No other changes.
Is someone out there with the same problems? Is it possible to downgrade just a single binding?

**Comment by J-N-K** [comment]
It's an issue with changed API on Tesla's side, I didn't upgrade and see the same issue.

**Comment by lsiepel** [comment]
While working on a [different PR](https://github.com/openhab/openhab-addons/pull/17348) that also has problems with a hard to reproduce / debug `EOFException` it was discovered this is caused by some transient network issue that cause the httpclient to fail.
Because the issue is fixed with a new httpclient (stop/start) i'm pretty sure it is the same issue. I guess there are multiple variants possible. Give me some time to create a PR. 

Would be nice if any of you who experience this issue are available to test an updated jar.

**Comment by hlemmens** [comment]
Happy to test, but with the actual binding is not possible to send charging commands to the new Tesla's. Tesla has again changed the protocol for new cars.

**Comment by lsiepel** [comment]
> ctual binding is not possible to send charging commands to the new Tesla's. Tesla has again changed the protocol for new cars.

Do you have any details on the change and to what cars / versions / firmware etc it relates?

**Comment by hlemmens** [comment]
You can find some information and an implementation example here: https://github.com/teslamotors/vehicle-command/blob/main/README.md

And this is the Tesla documentation on this issue: https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api

Hope this can help. I didn't succeed to make a working python program to send controls to my Tesla.

**Comment by lsiepel** [comment]
> You can find some information and an implementation example here: https://github.com/teslamotors/vehicle-command/blob/main/README.md
> 
> And this is the Tesla documentation on this issue: https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api
> 
> Hope this can help. I didn't succeed to make a working python program to send controls to my Tesla.

As i don't own a Tesla, i can only stick to small bugfixes. For an extensive change like this it is better that someone else picks this up.
